Adding SDR to NetworkAPI

dkeeney opened the pull request:
This is a very early hack up of this code.  Will not compile.
Primarally Types.hpp, BasicType.hpp/cpp, and ArrayBase.hpp/cpp.  Have not started to look at the Array class yet.

dkeeney commented:
There are more than expected complications in implementing the NTA_BasicTypes_Sparse data type.  We might want to backout my hacks and just drop the sparse array into an SDR before sending it over the Links.

dkeeney commented:
After I get these routines looking reasonable, I can make some minor mods to the Array class to support it.  Then look at the Link class to see how to take advantage of the new features.  Then we can look at each of the built-in regions to set them up so they can take advantage of the SDR class.

dkeeney commented:
@ctrl-z-9000-times sorry about the ugly code...its not as pretty as your SDR classes.  I think I can clean it up some with Lambda expressions and stop using C functions :)

ctrl-z-9000-times commented:
> Then we can look at each of the built-in regions to set them up so they can take advantage of the SDR class.

Eventually, I'd like to setup some of the algorithms to use SDR's.  Would this be useful for making the engine use SDR's too?

dkeeney commented:
By adding SDR to the Array class, it can be passed via the Link to the Region implementation.  That in turn could pass the SDR directly to the underlining algorithm in the  compute( ) call.  That is how we make the engine use the SDRs.

The trick is that each Region implementation (a subclass of RegionImpl class) contains a static data structure called the Spec that defines what the region expects for input and output buffers (as well as all parameters). VectorFileSensor.cpp is a C++ example. RegionSP.py and RegionTM.py are examples of python region implementations.  For Region implementations that can handle an SDR we set the second argument in the outputs and inputs of the Spec to specify NTA_BasicType_SDR.  When the Link object initializes, it will create Array objects for each input and output which are initialized with empty buffers (with dimensions) of the type specified in the Spec.  When the engine is told to run, these buffers are presented to the Region implementations to be passed to the algorithms.  The input buffer as input data and the output buffer to be filled in with the results.  The Link object then makes sure that the next recipient of the buffer gets the data.  If I add the data conversion capability to the Link interface then data should flow from region to region smoothly regardless of the type of data that a Region needs.

ctrl-z-9000-times commented:
>  If I add the data conversion capability to the Link interface

So the link has on one end an SDR, and on the other end a data buffer.  Does the link know if the data buffer is sparse or dense?

It will need to know this, in order to set/get the right thing from the SDR which contains both sparse & dense.

dkeeney commented:
> Does the link know if the data buffer is sparse or dense?

Currently only dense buffers exist.  If I add the NTA_BasicType_Sparse type then we can pass sparse data. The type tells me what format it is in.  @breznak added a separate 'sparse' flag but I don't think that any regions actually used that yet.  It would be replaced with NTA_BasicType_Sparse type.

ctrl-z-9000-times commented:
Thanks I understand now!

dkeeney commented:
Ran into a minor problem... I have a need for a const version of Array.getBuffer().  Specifically, I call getBuffer() in a operator== overload.  (same sort of thing with the operator<< overload.)
```
bool operator==(const ArrayBase &lhs, const ArrayBase &rhs) {
  if (lhs.getType() != rhs.getType() || lhs.getCount() != rhs.getCount())
    return false;
  if (lhs.getCount() == 0)
    return true;
  return (std::memcmp(lhs.getBuffer(), rhs.getBuffer(),
                      lhs.getCount() * BasicType::getSize(lhs.getType())) == 0);
}
```
I need to replace the memcmp( ), separate issue.
The problem is that if the Array contains an SCR I end up effectively doing this:
`       return getSDR()->getDense().data();`  in getBuffer( ).
But getDense( ) is not const. It modifies the SDR so not sure it could be const.
Any ideas?

ctrl-z-9000-times commented:
SDR by design is non constant, b/c it caches all of the format conversations.  There is a private constant method to get the data, but i would avoid using it if possible.  If its not possible then i can open it up as a public method (with an underscore in its name to discourage ppl from using it).

ctrl-z-9000-times commented:
IIRC the sdr has an operator overload for == and it is non constant.  It works in most situations, even if its not what ppl expect...

I think the unit tests needed to explicitly write == b/c the test macro casts everything to const. ASSERT_TRUE(A==B) instead of ASSERT_EQ(A,B)

dkeeney commented:
I will try this as non-const.  Not ideal but perhaps I can get it to work.

Another problem.  ArrayBase.h includes Sdr.h (so it can define SDR* getSDR( ),  Sdr.h includes Array.h which is a subclass of ArrayBase.h so Array cannot resolve ArrayBase.

You need Array so you can add an Array to the SDR.  If you are using Array then this must be within the NetworkAPI and the NetworkAPI is creating the SDR's inside the Array.  

`    void setFlatSparse( const ArrayBase &value ) {`
1) ArrayBase now cannot be const.
2) ArrayBase contains dense arrays rather than flat sparse.
Maybe we should just drop this function?

dkeeney commented:
This is also in `    void setDense( const ArrayBase &value )  { `
If you really need to load an Array value:
```
Real32* ptr = (Real32*)A.getBuffer( );  // assuming that is the type
scr.setDense(ptr);
```
The type and length of the buffer probably needs to be checked anyway.

dkeeney commented:
Oh. you would need a conversion too.
```
Real32* ptr = (Real32*)A.getBuffer( );  // assuming that is the type
        BasicType::convertArray(
            getDense().data(), NTA_BasicType_Byte,
            A.getBuffer(), A.getType(),
            A.getCount());
```

dkeeney commented:
Not allowing SCR to be const has another consequence...
ArrayBase has a save( ) which should be a const function because it overrides save( ) declaration in Serializable class.  But save() must also get at the SDR in order to access the data so it cannot be const.

So, I would have to change all save( ) functions to be non-const everywhere.  Hmmm, this is harder to resolve.

ctrl-z-9000-times commented:
I thought i fixed that... SDR.save should accept a constant SDR.  If it doesnt then tell me and i can & will fix it

dkeeney commented:
Found a workaround using keyword mutable.
I will code this up and see how it looks.

dkeeney commented:
No, save( ) in SDR was not the problem, its the save( ) in ArrayBase which must access the SDR.
So far using mutable is solving the problem. I am trying to compile it call up.

ctrl-z-9000-times commented:
> `mutable`

I didn't know that C++ had this feature...  This should make the SDR easier to work with and maybe I can simplify the implementation using this.  Thanks!

dkeeney commented:
The mutable keyword is good for fixing things like this BUT it does come at a price. With this set the compiler no longer check for const conditions on those variables and classes that contain them and it is possible to inadvertently allow modification when it was not intended.

dkeeney commented:
@ctrl-z-9000-times  
As I was working with the SDR class I found that I was having to make special provision for getting the dimensions into SDR.  The Link classes do provide facility for managing dimensions being manually assigned to either a region output or to a link. But they do not require that the dimensions be square. The Dimensions class includes the facilities for accessing a flat dense array with any set of dimensions you might want to use.

You have repeatedly stated that the dimensions must be square.  I always have the flattened dense size of the array, or element count, but not always the dimensions. I got to thinking about what that means.  If a two dimensional SDR must be square than the dimensions must be ceil( sqrt(count)).
And for a three dimensional SDR, it must be ceil(cbrt(count)).  They can have no other values.

Since dimensions are nothing more than a projection on the same set of data, why can't the SDR class always be created with a single dimension (i.e. the count) and that the SDR can calculate the 2nd or 3rd dimensions when someone wants to view the data in such a way?  In the same way, the sparse data caches could also be created on-the-fly when needed from calculated dimensions.

breznak commented:
> You have repeatedly stated that the dimensions must be square. I always have the flattened dense size of the array, or element count, but not always the dimensions. I got to thinking about what that means. If a two dimensional SDR must be square than the dimensions must be ceil( sqrt(count)).
And for a three dimensional SDR, it must be ceil(cbrt(count)). They can have no other values. 

Why is this? Does it mean if I have a SDR (size=100), dimensions must be 10x10? 

> Since dimensions are nothing more than a projection on the same set of data, why can't the SDR class always be created with a single dimension (i.e. the count) and that the SDR can calculate the 2nd or 3rd dimensions when someone wants to view the data in such a way? In the same way, the sparse data caches could also be created on-the-fly when needed from calculated dimensions. 

Like the SDR_Proxy does. By default, SDR stores in a "flat sparse", so it is a 1D UInt vector.

ctrl-z-9000-times commented:
> You have repeatedly stated that the dimensions must be square.

I'm sorry I must have made a mistake, that's wrong.
The dimensions must be *rectangular*, not square.

Example: if you have an SDR with size 100, it could have dimensions (2 x 5 x 10)

ctrl-z-9000-times commented:
> the SDR can calculate the 2nd or 3rd dimensions when someone wants to view the data in such a way?

To do this you will need to use the SDR_Proxy class, which does these conversions automatically. For example:
```
SDR A( dimensions = {1000} )
SDR_Proxy B( A, dimensions = {20, 10, 5} )
A.setData(...) <- input data has dimensions {1000}
B.getSparse()  -> output data has dimensions {20 x 10 x 5}
```

dkeeney commented:
Oh, big difference.
So I cannot just compute dimensions.  But that means I CAN use the
rectangular dimensions that are managed by the Link (the Dimension class).

So what happens if there is only one dimension given?


On Thu, Jan 31, 2019 at 7:36 AM David McDougall <notifications@github.com>
wrote:

> You have repeatedly stated that the dimensions must be square.
>
> I'm sorry I must have made a mistake, that's wrong.
> The dimensions must be *rectangular*, not square.
>
> Example: if you have an SDR with size 100, it could have dimensions (2 x 5
> x 10)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/htm-community/nupic.cpp/pull/234#issuecomment-459389085>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBa_25eBxSzTj4mLoKd5NYnraf8V6J3ks5vIw1lgaJpZM4aVAiH>
> .
>

dkeeney commented:
ok, so what I will be doing is creating the SDR with the rectangular dimensions if they were provided by the user and otherwise I will create it with single dimensions.

ctrl-z-9000-times commented:
> ok, so what I will be doing is creating the SDR with the rectangular dimensions if they were provided by the user and otherwise I will create it with single dimensions.

That sounds good to me.  1-D SDRs should work just like multi-dimensional SDRs.

ctrl-z-9000-times commented:
> So what happens if there is only one dimension given?

I plan on added a convenience to the SDR class:
`SDR( UInt dimensions ) : SDR( {dimensions} ) {};`
Which will allow you to make 1-D SDR's without first needed to make a vector containing a single integer.  Currently, the caller needs to make this dimensions vector.

dkeeney commented:
@ctrl-z-9000-times I have a unit test case where I am checking various Array type conversions. After populating the Array containing an SDR by writing directly to the dense buffer I now want to get a sparse version.  I call getFlatSparse().  But it returns an empty vector.  Obviously I am using this wrong.  Can you look at it for me?  ArrayTest.cpp Line 582.
I will push what I have so you can see it.

ctrl-z-9000-times commented:
@dkeeney, 
You need to call a setter method.  The setter methods notify the SDR when its data has changed, and that it needs to do the format conversions.  

**Inplace:** If you call `sdr.setDense()` with the correct datatype (which you have since you're modifying the sdr's own data) then it swaps data with the argument vector instead of copying it.   The line: `sdr.setDense( sdr.getDense() )` will assign the current dense data inplace.

ctrl-z-9000-times commented:
Ok, I think I fixed somethings.  There are fewer errors.

Recommendation: As soon as an SDR is assigned to an Array, set `array.count_ = sdr.size`.  There is a lot of code here which appears to rely on count_ instead of calling the getCount() method.

dkeeney commented:
>  The line: sdr.setDense( sdr.getDense() ) will assign the current dense data inplace.

@ctrl-z-9000-times Thanks,
I think I understand what this is doing. But this is going to be difficult to get the code right so that callers of Array (or ArrayBase) do not need to be aware of the fact that its an SDR they are writing to.

What I was trying to do was allow the caller of Array to call getBuffer() and it will always return a pointer to the beginning of the data buffer, regardless of the type.  That caller could read from or write to that raw buffer at will.  The problem is that I never now when the caller is finished using the pointer so that setDense() can be called.

But so much of the code is based on using the buffer pointer to access the raw data.  Changing that would require a major re-design.  What if I add a call to setDense( getDense()) whenever the caller makes a call  to getSDR( ).   If there were no changes prior, I assume that would do nothing, right?

ctrl-z-9000-times commented:
> What if I add a call to setDense( getDense()) whenever the caller makes a call to getSDR( ). If there were no changes prior, I assume that would do nothing, right?

Calling `setDense( getDense())` makes the SDR discard its current sparse & flatSparse data.  It also causes the SDR to call its callbacks, which will notify SDR_Metrics  that there is new data available for it to measure.

ctrl-z-9000-times commented:
> The problem is that I never know when the caller is finished using the pointer so that setDense() can be called.

Maybe you could assume that each region will make one assignment, and call sdr.setDense() after every region.compute(sdr)

dkeeney commented:
> call sdr.setDense() after every region.compute(sdr)

There is a problem with scope. Most of the conversions and buffer copies is happening within the links prior to each execute iteration. This is before any regions are called to execute.  Of course, the Array buffers are also accessed within the Region Implantation.

At the level of the Array class, it knows about SDR vs raw buffers, but it knows nothing about Links or regions.  It is just a very low level wrapper class used to pass data around.  

Outside of Array, the callers should not know about the SDR ... unless they are executing within a Region implementation that has declared that it wants an SDR by saying so in the Spec.  Then it can call Array.getSDR() and then from then on it does not need the Array class.  We will assume that this code knows to cleanup the cache before exiting the region.  

Here is what I could do:
Just before the engine starts calling region implementations for an iteration, the current code walks the links and prepares the data. It moves the source buffer to the destination buffer on each link.  Most of the time there is no conversion or data copies but if the types do not match, a delayed propogation, or a fan-in there will be a conversion/copy.  At that point I could have the Link, call a new function on the Array which says .. if this is an SDR, call set_dense(get_dense()).

So what shall I call it?   Synch( )?  Prepair( )?  Cache( )?  Maybe just PrepareCache( )
Its hard to pick a name that describes what this is doing in the context of the Array. 
I will use  PrepareCache( ) until a better name can be devised.

dkeeney commented:
@ctrl-z-9000-times I have made some changes to Sdr.h  I would like you to take a look and see what you think. Perhaps you have a better way of addressing the problems.
- removed functions that accepted an Array as input.  My thinking is that the Array should contain the SDR rather than the other way around.  This solved the problem of including Array.h in the wrong order.
- Adding mutable to several of the protected fields.  This allows the SDR to be declared const when appropriate and yet allowing the cache to be updated with getDense( ) when called within the context of a function that is const.  So there is now a const version of getDense( ) and getFlatSparse( ).
- mismatching type conversion - possible loss of data; --- In setSparse(vector), the loop that adds sparse data elements into its internal vectors calls sparse[dim].assign(begin, end) but this implys a type conversion that MSVC does not like.  I replaced it with a simple loop so that a cast could be added to tell the compiler that it is ok.
- another complication. The NetworkAPI expects to be able to allocate a buffer of zero size.  This would be a buffer which is unused or for which data is not provided as opposed to a buffer that was not initialized.  SDR does not allow zero length dimensions.  I don't have a fix for this. Perhaps setting dimensions to { 1 } and then tell it to truncate length to 0. But I don't see a way to set a capacity larger than its actual value.  I would like an SDR that does not do much but has zero length vectors and will return 0 for size.  It should complain on any attempt to add data beyond capacity.

I will push what I have.

dkeeney commented:
> NetworkAPI expects to be able to allocate a buffer of zero size.

For the zero_sized buffer, I am going to try to let ArrayBase detect and handle the condition.  I will see how this goes.

ctrl-z-9000-times commented:
> removed functions that accepted an Array as input. 

Good.

> Mutable

I have not looked at the code but i like this.  Im considering a new PR to add mutable and const throughout the sdr,  this change will also enable minor cleanup in sdr.  You should go ahead and do this as needed

> mismatching type conversion 

Thank you for making windows work!  Do what you need to do.

> SDR(0)

Yea this is a problem.  A zero sized sdr is a degenerate case, but its useful for making placeholders and such

dkeeney commented:
I have something that runs on MSVC now and passes the Unit tests on Windows.

But I am not entirely happy with it.  The functions available in the Array and  ArrayBase seem to cover most requirements but they seem to me to be haphazard and hodgepodge.   Not a well thought out set of function calls.
```
//
// POPULATING:
//   constructor A(type)                   - Empty buffer of specified type.
//   constructor A(type, buff, size)       - Buffer filled from C array
//   constructor A(sdr)                    - Buffer initialized from SDR
//   B = A;                                - Shallow copy, B has same, type, buffer, size
//   A.share(B)                            - Shallow copy, B has same, type, buffer, size
//   A.copyfrom(type,buff, size)           - fills A with buff[size], sets type.
//   B = A.copy()                          - Creates B as a copy of A.
//   A.populate(vector)                    - fills A, with conversion, A retains type.
//   A.fromDenseVector(vector)             - fills A, A becomes type of vector.
//   A.fromSparseVector(vector, dimensions)- fills A with sparse data, A becomes SDR type
//   A.zeroBuffer()                        - fills A with 0's, A retains type and size.
//
//
// ACCESSING:
//   vector = A.asVector()        - creates vector and populates it from A
//   B = A.get_as(type)           - creates B with conversion, to specified type.
//   A.convertInto(B, offset)     - fills B in-place with conversion, B retains type.
//   B = A.subset(offset, length) - fills B with subset of A, same type.
```

For example:
```
B = A;
A.share(B);   
```
Perform the same function but with A.share(B), B already exists.  Probably should get rid of A.share(B).

Perhaps some sort of uniform way of indicating in the function name if  type conversion is performed.
Perhaps add a constructor that populates from a vector.
Are there other things that I should do?

dkeeney commented:
Oh, I am missing the set from ArrayBase
```
//
// FROM ArrayBase:
// A.AllocateBuffer(count)           -- single dimension buffer allocation
// A.AllocateBuffer(dimensions)      -- multi dimensional buffer allocation (for SDR)
// A.getBuffer()                     -- returns a void* pointer to beginning of buffer.
// A.setBuffer(ptr, count)           -- set un-owned buffer
// A.setBuffer(sdr)                  -- set un-owned SDR
// A.zeroBuffer()                    -- fills A with 0's, A retains type and size.
// A.releaseBuffer()                 -- free everything (if owned)
// A.getSDR()                        -- get pointer to enclosed SDR
// A.getBufferSize( )                -- size of buffer in bytes
// A.getMaxElementsCount()           -- capacity in number of elements 
// A.setCount(count)                 -- truncate buffer size to this length, keeping capacity.   (not SDR)
// A.getType()                       -- return NTA_BasicType of buffer
// A.has_buffer()                    -- returns true if a buffer has been allocated.
// A.is_instance(B)                  -- returns true if B is shared with A
// A.RefreshCache()                  -- tells SDR to update cache
// A.save(stream)                    -- serialize
// A.load(stream)                    -- deserialize
```

ctrl-z-9000-times commented on src/nupic/engine/Link.cpp:
```diff
@@ -40,9 +40,6 @@
 
-// Represents 'zero' scalar value used to compare Input/Output buffer contents
-// for non-zero values
-const static Real64 ZERO_VALUE = 0;
```
lol

ctrl-z-9000-times commented:
Question: is `NTA_BasicTypes_Sparse` used with SDR's?  Or does is it replaced by SDR?  I thought it was replaced but then I saw some references to it at:
- src/nupic/types/Types.hpp line 228
-  BasicType.cpp

ctrl-z-9000-times commented on src/nupic/engine/Input.cpp:
```diff
@@ -460,35 +465,27 @@ void Input::initialize() {
   // Later we may optimize with the zeroCopyEnabled_ flag but
-  // for now we always allocate our own buffer.
+  // for now we always allocate our own input buffer.
+  // Create the Input buffer.
```
Questions, out of curiosity:  I see some changes to fan-in/fan-out, does it copy or work inplace?

dkeeney replied:
>  I see some changes to fan-in/fan-out, does it copy or work inplace?

For fan-in there must be a copy because each output is being written to a different offset in the same input buffer.  Otherwise, all connections are non-copy unless the data type is different OR the link uses a splitter map to remap the bits in the buffer.  Originally all links required a copy.

ctrl-z-9000-times commented on src/nupic/types/BasicType.cpp:
```diff
@@ -212,8 +225,19 @@ static void cpyarray(void *toPtr, const void *fromPtr, size_t count, F minVal, F
 }
 
+template <typename T>
+static void cpyIntoSDR(Byte *toPtr, const T *fromPtr, size_t count) {
```
Switch statement is now 446 lines long  :)

dkeeney replied:
> Switch statement is now 446 lines long :)

Actually this is a tree of switch statements. No one switch is more than 12 elements.
I could not think of a better way to do this. It cannot be templated because we are not switching on C++ type.  Perhaps with 12 polymorphous classes.   It may be ugly but it is readable.

ctrl-z-9000-times commented on src/nupic/ntypes/ArrayBase.hpp:
```diff
@@ -186,8 +238,8 @@ namespace nupic
-  bool operator==(const ArrayBase &lhs, const ArrayBase &rhs);
-  inline bool operator!=(const ArrayBase &lhs, const ArrayBase &rhs) {return !(lhs == rhs);}
+  bool operator==(ArrayBase &lhs, ArrayBase &rhs);
+  inline bool operator!=(ArrayBase &lhs, ArrayBase &rhs) {return !(lhs == rhs);}
```
Is this a mistake or intentional?  I though that you made `const ArrayBase` possible.

dkeeney replied:
>  I though that you made `const ArrayBase` possible.

Oh, good catch.  A left-over when I was trying to make ArrayBase work with non-const SDR.  I will put the const back in.

ctrl-z-9000-times commented on src/nupic/ntypes/ArrayRef.hpp:
```diff
@@ -62,12 +62,25 @@
 namespace nupic {
+//typedef const Array ArrayRef;
+
+#if 0
```
Can this file be deleted?

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -148,11 +146,21 @@ void ArrayBase::zeroBuffer() {
+void ArrayBase::setBuffer(SDR &sdr) {
+  type_ = NTA_BasicType_SDR;
+  buffer_ = std::shared_ptr<char>((char *)&sdr, nonDeleter());
+  count_ = sdr.size;
```
you can call `setBuffer(std::shared_ptr<char>((char *)&sdr, nonDeleter()), sdr.size)`

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -210,9 +229,12 @@ size_t ArrayBase::getCount() const {
+  if (has_buffer()) {
+    if (type_ == NTA_BasicType_SDR)
+      return getCount();
+    return capacity_ / BasicType::getSize(type_);
```
NTA_ASSERT(capacity_ % getSize(type_) == 0 ) << "Would lose precision by rounding"

breznak replied:
This ^

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -221,10 +243,9 @@ size_t ArrayBase::getMaxElementsCount() const {
  */
 void ArrayBase::setCount(size_t count) {
   NTA_CHECK(type_ != NTA_BasicType_SDR) << "Operation not valid for SDR";
-  NTA_ASSERT(count <= capacity_ / BasicType::getSize(type_))
```
NTA_ASSERT capacity_ is divisible by getSize() w/o reminder

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
+/**
+ * Return true if a buffer has been allocated.
+ */
+bool ArrayBase::has_buffer() const { return (buffer_.get() != nullptr); }
```
nit, stick to camelCase next time, please

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
+ * them in order so that the largest index is last.
+ *
+ * Be careful when using this with destination of SDR...it will remove
+ * dimensions if buffer is  not big enough.
```
can we rather throw an err in that case?

breznak replied:
this ^

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
-      }
-    //getSDR()->setFlatSparseInplace();
-    }
+    NonZeroT<bool>(a, this, offset);
```
instead of all these switch-case, can you do `NonZeroT<type_>(a, this, offset)` ?

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
-      for (size_t i = 0u; i < count_; i++) {
-        *p1++ = *p2++;
-      }
+  case NTA_BasicType_SDR: {
```
and you might need to handle SDR, Sparse types before that (above)

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
-  switch (a.getType()) {
+  switch (a.getType()) { // to an Array of this type.
   case NTA_BasicType_Byte:
-    DenseT<Byte>(a, size, value, count_, offset);
```
ditto with the `DenseT<a.getType())`

breznak commented on src/nupic/ntypes/ArrayBase.hpp:
```diff
@@ -98,9 +103,10 @@ namespace nupic
 
     /**
      * Ask ArrayBase to allocate its buffer
+     * NOTE: for NTA_BasicTypeSparse this sets the size of the dense buffer is describes.
```
nit _Sparse

breznak commented on src/nupic/ntypes/ArrayBase.hpp:
```diff
@@ -126,33 +132,70 @@ namespace nupic
+    /**
+     * max number of elements this buffer can hold (capacity)
+     */
 	  size_t getMaxElementsCount() const;
```
nit whitespace

dkeeney replied:
> nit whitespace

? it aligns on my screen.  I will see if there is a tab in there.  Actually I think I have tabs all converted to spaces.

breznak commented on src/nupic/ntypes/ArrayBase.hpp:
```diff
@@ -126,33 +132,70 @@ namespace nupic
+    void inline RefreshCache() {
+      if (type_ == NTA_BasicType_SDR) {
+        SDR *sdr = getSDR();
+        sdr->setDense(sdr->getDense());
```
add explanation why, what does this method do

breznak commented on src/nupic/types/Types.hpp:
```diff
@@ -42,8 +42,12 @@ namespace nupic {
 /**
- * Represents a 8-bit byte.
+ * Represents a signed 8-bit byte.
+ * (note that std::byte is unsigned char.)
```
and we do need signed?

dkeeney replied:
> nupic::Byte: and we do need signed?

No, but I have verified that it does compile with both signed and unsigned.  What I am trying to point out is that this is an integer value NOT a character.  Unfortunately we don't have a platform independent way to tell the compiler that.  Will still have problems if used with I/O on windows.

breznak replied:
> No, but I have verified that it does compile with both signed and unsigned. 

I recall there were some problems with Byte = std::byte ? If not, wouldn't unsigned be better for this?

dkeeney replied:
> I recall there were some problems with Byte = std::byte ?

std::byte would be ideal because this is what this type is intended for. An 8bit integer and not a character. However, it is not in the standard until C++14.   So we should leave it as it always has been, a char (which is a signed char).

breznak commented on src/test/unit/engine/InputTest.cpp:
```diff
@@ -119,20 +119,17 @@ TEST(InputTest, SplitterMap) {
+    ArrayBase *ao1 = &(out1->getData());
+    Real64 *ptr = (Real64*)ao1->getBuffer();
+    for (size_t i = 0; i < ao1->getCount(); i++)
+      *ptr++ = 10;
```
nit, imho the array index[i] was more readable

dkeeney replied:
> nit, imho the array index[i] was more readable

Depends if you come from the C world...  I will change it.

dkeeney commented:
> Question: is `NTA_BasicTypes_Sparse` used with SDR's?

Both NTA_BasicTypes_Sparse and NTA_BasicTypes_SDR are new types.  The NTA_BasicTypes_Sparse is used to replace is_sparse flags  scattered around in the link, input, and output objects.  NTA_BasicTypes_Sparse is a request from @breznak.  It is indeed a duplicate of the capability of the SDR.

breznak commented:
> NTA_BasicTypes_Sparse is a request from @breznak. It is indeed a duplicate of the capability of the SDR. 

I'm sorry for the confusion, with the current state of the PR, I'm also for Sparse being obsoleted by SDR. Unless there is, if we have them (SDR, Sparse) both now, some advantage to Sparse?

dkeeney commented:
>  IMO the NetworkAPI is more complicated than the problems which it solves.

Perhaps, but my hope that with the right documentation, the user will find it less complicated to do applications that are non-trivial.

dkeeney commented:
>  is the "Sparse" type so far used? If not, we could remove it and just use new SDR for that purposes.

@breznak the Sparse type was added at your request.  But I agree that the SDR is a better way to handle this so I will remove it.

breznak commented:
>  the Sparse type was added at your request. But I agree that the SDR is a better way to handle this so I will remove it. 

:+1: 
Sorry I didn't see that earlier. Sparse type was added to substitute wrong use of 'sparse' flag. Meanwhile we added SDR which is a Sparse container.

breznak commented:
> > IMO the NetworkAPI is more complicated than the problems which it solves.

By NetworkAPI you mean just the types, Links, Policies, etc., or even the Regions? 

> Perhaps, but my hope that with the right documentation, the user will find it less complicated to do applications that are non-trivial.

Agree. I also (used to) find NetworkAPI more complicated than using directly the core (SP, TM) classes. But as Matt said, it is well documented, and shields the implementation away.

dkeeney commented on src/nupic/engine/Input.cpp:
```diff
@@ -537,6 +534,7 @@ const std::vector<std::vector<size_t>> &Input::getSplitterMap() const {
 
 template <typename T>
 void Input::getInputForNode(size_t nodeIndex, std::vector<T> &input) const {
+  // NOTE: nodeIndex is always 1 because we no longer support nodes.
```
> // NOTE: nodeIndex is always 1 because we no longer support nodes.

This was a comment explaining what I found out but I am not so sure now.  Here is what I am finding:

- If you look at all of the regions implemented in python you will see that singleNodeOnly = true in the Spec structure for all of the region implementations.  The only exceptions that I found were in C++, namely TestNode.cpp, VectorFileEffector, and VectorFileSensor.   
- What that flag means is to disable the ability to set dimensions on regions or links.  The 'node' in the comment refers to the first dimension or row of at least a two dimensional array for an input buffer.  
- The getInputForNode( ) function takes the input buffer and re-maps the bits using the splitter map, calling once for each row in the input buffer.  Without dimensions there is only one row and the splitter maps are not very useful...not sure they can even be created.
- The splitter maps are the product of the LinkPolicy and its subclasses TestFanIn2LinkPolicy and UniformLinkPolicy.  In my opinion a link policy is too complicated to setup and there is almost no documentation on how to do it.
- So at the moment, there can be multple 'nodes' if dimensions are defined and splitter maps do work when calling the getInputForNode( ) function.

What I wanted to do was remove dimensions and all LinkPolicy logic.  This would leave the feature of output buffers to be connected to input buffers with no remapping of bits.  Fan-In would just append output buffers side-by-side in the input buffer.  Fan-out would not be affected.  Propogation delay would not be affected.  I actually coded this all up about a year ago.

However, with the SDR it seems that dimensions are back.  So I am hesitant to rip out dimensions from the Links, just in case we need dimensions for the SDRs.  So, I made sure that LinkPolicy and splitter maps also still work with it.

The question is...(and I have asked this before)... should I rip dimensions, LinkPolicy, and splitter maps out of Links and make everything much simpler?  When I asked the question on the forum I did not get a definite answer because I don't think people even know what that means.

So, until told otherwise I will leave them all working... and I will remove that Note.

breznak replied:
This is a useful memo, thank you! 

> What I wanted to do was remove dimensions and all LinkPolicy logic. This would leave the feature of output buffers to be connected to input buffers with no remapping of bits. Fan-In would just append output buffers side-by-side in the input buffer. Fan-out would not be affected. Propogation delay would not be affected. I actually coded this all up about a year ago. 

I would like to see such simplification in the API. @ctrl-z-9000-times also complained it's overcomplicated. 

> However, with the SDR it seems that dimensions are back. So I am hesitant to rip out dimensions from the Links, just in case we need dimensions for the SDRs. So, I made sure that LinkPolicy and splitter maps also still work with it. 

COuld be always use 1D view of SDRs for that? Or, I'll reiterate on the issue of Sparse vs SDR types, Sparse is not affected by this, right? So do it (Sparse-SDR one becomes obsoleted) the other way around and keep Sparse? (But that would lose some nice benefits of using SDR, as no-copy?)

dkeeney replied:
> I'll reiterate on the issue of Sparse vs SDR types, Sparse is not affected by this, right?

Not sure what you are asking.  The splitter maps do not make since unless the data is the Dense format of SDR data (with the nupic definition of SDR).  Other types of data such as reset, ids, or numeric data that is not zero/non-zero would not use the splitter maps.  Sparse data format (just the indexes) also would not make since with a splitter map unless it was converted to dense format.  However by using the SDR class it will do the right thing using the Dense data in the class.

dkeeney commented on src/nupic/ntypes/Array.hpp:
```diff
@@ -168,7 +208,7 @@ class Array : public ArrayBase {
    * will make a full copy of the buffer and becomes the buffer owner.
    */
-  Array copy() const {
+  Array copy() {
```
> why not const anymore?

I can put that back in as well.

dkeeney commented on src/nupic/ntypes/Array.hpp:
```diff
@@ -195,92 +236,136 @@ class Array : public ArrayBase {
+        UInt32 *Destptr = (UInt32 *)getBuffer();
+        for (UInt32 i = 0u; i < getCount(); i++) {
+          if (v[i])
+            Destptr[j++] = (UInt32)i;
```
> otherwise this would crash, ie for T=string

I don't think it would compile for a string.  It must be one of the valid temples. See BasicType.cpp line 100.

dkeeney commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -38,31 +38,32 @@ namespace nupic {
-    }
+  allocateBuffer(count);
+  if (has_buffer()) {
+    std::memcpy((char *)getBuffer(), (char *)buffer,
```
> can this be part of allocateBuffer() already?

don't understand your question.  allocateBuffer() just creates the space.  There are lots of ways in which it might be populated.

breznak replied:
I saw it later, I had thought the memcpy part was alwasy the same.

dkeeney commented:
So, the final result of all of this review is mostly to remove NTA_BasicType_Sparse.
I will continue to retain dimensions (and LinkPolicy, splitter maps) in the Links until told otherwise.

breznak commented:
> I will continue to retain dimensions (and LinkPolicy, splitter maps) in the Links until told otherwise.

We can revisit that in another PR/issue, I think the cleanup would be :+1: esp since you've coded that already somehow

dkeeney commented:
Here is my new function list 
```
// POPULATING:
//   constructor A(type)              - Empty buffer of specified type.
//   constructor A(type, buff, size)  - Buffer filled from C array
//   constructor A(sdr)               - Buffer initialized from SDR
//   constructor A(vector)            - Buffer initialized from vector
//   B = A;                           - Shallow copy, B has same, type, buffer, size
//   B = A.copy()                     - Creates B as a copy of A.
//   A.populate(vector)               - fills A from vector, with conversion, A retains type.
//
//
// ACCESSING:
//   vector = A.asVector()            - creates vector and populates it from A
//   B = A.get_as(type)               - creates B with conversion, to specified type.
//   A.convertInto(B, offset, maxsize)- fills B with conversion at offset, B retains type.
//   B = A.subset(offset, length)     - fills B with subset of A, same type.
//
//
// FROM ArrayBase:
// A.AllocateBuffer(count)           -- single dimension buffer allocation
// A.AllocateBuffer(dimensions)      -- multi dimensional buffer allocation (for SDR)
// A.getBuffer()                     -- returns a void* pointer to beginning of buffer.
// A.setBuffer(ptr, count)           -- set un-owned buffer
// A.setBuffer(sdr)                  -- set un-owned SDR
// A.zeroBuffer()                    -- fills A with 0's, A retains type and size.
// A.releaseBuffer()                 -- free everything (if owned)
// A.getSDR()                        -- get pointer to enclosed SDR
// A.getBufferSize( )                -- size of buffer in bytes
// A.getMaxElementsCount()           -- capacity in number of elements 
// A.setCount(count)                 -- truncate buffer size to this length, keeping capacity.   (not SDR)
// A.getType()                       -- return NTA_BasicType of buffer
// A.has_buffer()                    -- returns true if a buffer has been allocated.
// A.is_instance(B)                  -- returns true if B is shared with A
// A.RefreshCache()                  -- tells SDR to update cache
// A.save(stream)                    -- serialize
// A.load(stream)                    -- deserialize
// cout << A << std::endl;           -- stream out
// cin  >> A;                        -- stream in
// A == B                            -- equals overload
// A != B                            -- not equals overload

```

- I need to point out that I cannot do the subset( ) function if the type is SDR.  
- Otherwise it is much cleaner without the methods needed to support Sparse as a base type.  Now for all sparse things, just get the SDR object and use its functions.
- It might be cool to have an operator+ overload to append buffers side-by-side but not required.

dkeeney commented:
Ok, NTA_BasicType_Sparse is gone.   Much easier to remove than to put in  :)
Lets see if it builds...  

Oh, and I figured out how to prevent the duplicate appveyor builds :)

breznak commented:
https://travis-ci.org/htm-community/nupic.cpp/builds/489759713#L922 

Still seems to be some ArrayRef leftovers in bindings/py

dkeeney commented:
I hope the build works this time.
Sorry I had to take the afternoon off.  Had a network installation to do.  You know...running wires, putting on connectors, etc. :)

breznak commented:
> NTA_ASSERT(capacity_ % getSize(type_) == 0 ) << "Would lose precision by rounding" 

hmm, it did not add to the prev conversation. Please just revisit the unresolved comments above

ctrl-z-9000-times commented:
This looks good to me.

dkeeney commented:
> NTA_ASSERT(capacity_ % getSize(type_) == 0 ) << "Would lose precision by rounding"

Could not find this.  capacity_ cannot be anything other than even type sized increments because that is how its created.  So such an assert would not be useful.

>  Be careful when using this with destination of SDR...it will remove dimensions if buffer is  not big enough.

I assume this is what you are pointing out.  Yes, the convertInto( ) function  tries to perform this conversion in-place with an offset.  But if the destination buffer is too small, it will automatically re-create the buffer before it starts the copy.   When it does the re-creation it only has 1D  information so it cannot extend the size of the array without loosing the dimensions.  Within the Links this should not happen because the Link checks sizes during initialization to make sure everything works as expected.  It is only when this function is used is some isolated situation the programmer should be aware of this.
